File: gods_graphical/menu.py
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum, auto

import pyray

from gods_online.room_code import Connection_State, join_room, start_hosting
from kitchen_table.config import tweak
from kitchen_table.rendering import color_from_tuple, draw_background


class Screen(Enum):
    MAIN = auto()
    ONLINE = auto()
    CREATING = auto()    # Host: displaying room code, waiting for joiner.
    JOINING = auto()     # Joiner: text input for room code.
    CONNECTING = auto()  # Joiner: connecting in progress.


@dataclass
class Menu_State:
    screen: Screen = Screen.MAIN
    text_input: str = ""
    connection: Connection_State | None = None
    error_message: str = ""


# --- Helpers ---

def _centered_x(width: int) -> int:
    return (tweak["window_width"] - width) // 2


def _draw_centered_text(text: str, y: int, font_size: int,
                        color: tuple = (255, 255, 255, 255)) -> None:
    w = pyray.measure_text(text, font_size)
    pyray.draw_text(text, _centered_x(w), y, font_size, pyray.Color(*color))


def _draw_button(text: str, y: int, width: int = 320, height: int = 58) -> bool:
    """Draw a horizontally-centered button and return True if clicked this frame."""
    x = _centered_x(width)
    mx, my = pyray.get_mouse_x(), pyray.get_mouse_y()
    hovered = x <= mx <= x + width and y <= my <= y + height
    color_key = "button_hover_color" if hovered else "button_color"
    bg = color_from_tuple(tweak[color_key])
    fg = color_from_tuple(tweak["button_text_color"])
    pyray.draw_rectangle_rounded(pyray.Rectangle(x, y, width, height), 0.3, 8, bg)
    tw = pyray.measure_text(text, 22)
    pyray.draw_text(text, x + (width - tw) // 2, y + (height - 22) // 2, 22, fg)
    return hovered and pyray.is_mouse_button_pressed(pyray.MouseButton.MOUSE_BUTTON_LEFT)


def _draw_text_input(label: str, text: str, y: int,
                     width: int = 380, height: int = 52) -> None:
    """Draw a centered, labeled text input box with a blinking cursor."""
    x = _centered_x(width)
    label_w = pyray.measure_text(label, 18)
    pyray.draw_text(label, _centered_x(label_w), y - 30, 18, pyray.Color(200, 200, 200, 255))
    pyray.draw_rectangle_rounded(
        pyray.Rectangle(x, y, width, height), 0.2, 8, pyray.Color(30, 30, 50, 220)
    )
    pyray.draw_rectangle_rounded_lines_ex(
        pyray.Rectangle(x, y, width, height), 0.2, 8, 2, pyray.Color(140, 140, 200, 255)
    )
    # Cursor blinks at 1 Hz.
    cursor = "_" if int(pyray.get_time() * 2) % 2 == 0 else " "
    pyray.draw_text(text + cursor, x + 12, y + (height - 24) // 2, 24,
                    pyray.Color(255, 255, 255, 255))


def _update_text_input(text: str, max_length: int = 16) -> str:
    """Consume all pending key presses and return the updated text."""
    char = pyray.get_char_pressed()
    while char:
        if len(text) < max_length and 32 <= char < 127:
            text += chr(char)
        char = pyray.get_char_pressed()
    if pyray.is_key_pressed(pyray.KeyboardKey.KEY_BACKSPACE) and text:
        text = text[:-1]
    return text


def _dots() -> str:
    """Animated ellipsis string cycling through 0-3 dots."""
    return "." * (int(pyray.get_time() * 2) % 4)


# --- Main menu entry point ---

def run_menu() -> tuple[str, dict]:
    """Display the game menu and return the user's game choice.

    Opens a Raylib window that is kept open after this function returns so that
    the game can continue rendering in the same window (play() skips init_window
    when the window is already ready).

    Returns (mode, params) where:
      - mode == "vs_ai"    → params == {}
      - mode == "online"   → params == {player_index, seed, sock, friend_addr}
    """
    W = tweak["window_width"]
    H = tweak["window_height"]

    pyray.set_config_flags(pyray.ConfigFlags.FLAG_WINDOW_HIGHDPI)
    pyray.init_window(W, H, "Gods")
    pyray.set_target_fps(tweak["target_fps"])

    state = Menu_State()
    center_y = H // 2

    while not pyray.window_should_close():
        # --- Per-frame text input ---
        if state.screen == Screen.JOINING:
            state.text_input = _update_text_input(state.text_input)
            if pyray.is_key_pressed(pyray.KeyboardKey.KEY_ENTER) and state.text_input:
                state.connection = join_room(state.text_input)
                state.screen = Screen.CONNECTING

        # --- Poll async connection result ---
        if state.connection is not None:
            if state.connection.ready:
                c = state.connection
                return "online", {
                    "player_index": c.player_index,
                    "seed":         c.seed,
                    "sock":         c.sock,
                    "friend_addr":  c.friend_addr,
                }
            if state.connection.error:
                state.error_message = state.connection.error
                state.connection = None
                state.screen = Screen.ONLINE

        # --- Draw ---
        pyray.begin_drawing()
        draw_background()

        if state.screen == Screen.MAIN:
            _draw_centered_text("GODS", center_y - 180, 90)

            if _draw_button("Play vs AI", center_y - 20):
                return "vs_ai", {}

            if _draw_button("Play Online", center_y + 60):
                state.screen = Screen.ONLINE

        elif state.screen == Screen.ONLINE:
            _draw_centered_text("PLAY ONLINE", 150, 54)

            if state.error_message:
                _draw_centered_text(
                    state.error_message, center_y - 120, 18, (255, 100, 100, 255)
                )

            if _draw_button("Create Game", center_y - 60):
                state.error_message = ""
                state.connection = start_hosting()
                state.screen = Screen.CREATING

            if _draw_button("Join Game", center_y + 20):
                state.error_message = ""
                state.text_input = ""
                state.screen = Screen.JOINING

            if _draw_button("Back", center_y + 120, width=180, height=46):
                state.error_message = ""
                state.screen = Screen.MAIN

        elif state.screen == Screen.CREATING:
            _draw_centered_text("CREATE GAME", 150, 54)

            if state.connection and state.connection.room_code:
                _draw_centered_text(
                    "Share this code with your friend:", center_y - 80, 20,
                    (200, 200, 200, 255)
                )
                # Room code displayed prominently in gold.
                _draw_centered_text(
                    state.connection.room_code, center_y - 30, 50, (255, 215, 0, 255)
                )
                _draw_centered_text(
                    f"Waiting for opponent{_dots()}", center_y + 50, 22,
                    (180, 180, 180, 255)
                )
            else:
                _draw_centered_text(
                    f"Getting your room code{_dots()}", center_y, 26,
                    (200, 200, 200, 255)
                )

            if _draw_button("Back", center_y + 150, width=180, height=46):
                state.connection = None
                state.screen = Screen.ONLINE

        elif state.screen == Screen.JOINING:
            _draw_centered_text("JOIN GAME", 150, 54)
            _draw_text_input("Enter room code:", state.text_input, center_y - 40)

            if _draw_button("Connect", center_y + 50) and state.text_input:
                state.connection = join_room(state.text_input)
                state.screen = Screen.CONNECTING

            if _draw_button("Back", center_y + 130, width=180, height=46):
                state.text_input = ""
                state.screen = Screen.ONLINE

        elif state.screen == Screen.CONNECTING:
            _draw_centered_text("JOIN GAME", 150, 54)
            _draw_centered_text(
                f"Connecting{_dots()}", center_y - 20, 30, (200, 200, 200, 255)
            )

            if _draw_button("Back", center_y + 100, width=180, height=46):
                state.connection = None
                state.screen = Screen.JOINING

        pyray.end_drawing()

    # User closed the window during the menu.
    pyray.close_window()
    raise SystemExit(0)

File: gods_graphical/menu.cpp
#include "menu.h"

#include <raylib.h>
#include <tabletop/config.h>

#ifdef __EMSCRIPTEN__
#include <emscripten.h>
#include <emscripten/html5.h>

Menu_Result run_menu(Input_Feed&) {
  InitWindow(tt::WINDOW_WIDTH, tt::WINDOW_HEIGHT, "Gods");
  // FLAG_WINDOW_HIGHDPI is not implemented on PLATFORM_WEB (GetWindowScaleDPI
  // returns {1,1}).  Resize the canvas pixel buffer to physical resolution and
  // pin the CSS size to logical dimensions so the game fills the viewport at
  // full Retina sharpness.  The per-frame projection fix in draw_background
  // maps the logical 1700×1000 coordinate space to the physical canvas.
  double dpr = emscripten_get_device_pixel_ratio();
  if (dpr > 1.0) {
    emscripten_set_canvas_element_size(
      "#canvas", (int)(tt::WINDOW_WIDTH * dpr), (int)(tt::WINDOW_HEIGHT * dpr)
    );
    EM_ASM(
      {
        var c          = document.getElementById('canvas');
        c.style.width  = $0 + 'px';
        c.style.height = $1 + 'px';
      },
      tt::WINDOW_WIDTH,
      tt::WINDOW_HEIGHT
    );
  }
  SetTargetFPS(tt::TARGET_FPS);
  return Menu_Result{};  // Default mode is VS_AI.
}

#else

#include <tabletop/rendering.h>
#include <tabletop/ui.h>

#include <algorithm>
#include <cstdlib>
#include <memory>

// Keeps the UDP socket alive for the process lifetime. online_lib hands us a
// shared_ptr; we park one copy here so main()/Online can hold a raw pointer
// without worrying about ownership. Confined to this file so the rest of
// gods_graphical only sees raw pointers.
static std::shared_ptr<UDP_Socket> s_socket_keep_alive;

enum class Screen { MAIN, ONLINE, CREATING, JOINING, CONNECTING };

struct Menu_State {
  Screen                            screen = Screen::MAIN;
  std::string                       text_input;
  std::shared_ptr<Connection_State> connection;
  std::string                       error_message;
};

static int centered_x(int width) { return (tt::WINDOW_WIDTH - width) / 2; }

static void draw_centered_text(
  const std::string& text, int y, int font_size, Color color
) {
  int w = text_width(text, font_size);
  render_text(text, (float)centered_x(w), (float)y, font_size, color);
}

static bool draw_button(
  const Input&       input,
  const std::string& text,
  int                y,
  int                width  = 320,
  int                height = 58
) {
  int  mx      = input.mouse_x;
  int  my      = input.mouse_y;
  int  x       = centered_x(width);
  bool hovered = (x <= mx && mx <= x + width && y <= my && my <= y + height);

  if (hovered) {
    DrawRectangleRounded(
      Rectangle{(float)x, (float)y, (float)width, (float)height},
      0.3f,
      8,
      ::Color{20, 20, 20, 100}
    );
  }

  int size = 30;
  int tw   = text_width(text, size);
  render_text(
    text,
    (float)(x + (width - tw) / 2),
    (float)(y + (height - size) / 2),
    size,
    Color{255, 255, 255, 255}
  );

  return hovered && input.left_pressed;
}

static void draw_text_input(
  const std::string& label,
  const std::string& text,
  int                y,
  int                width  = 380,
  int                height = 52
) {
  int x       = centered_x(width);
  int label_w = text_width(label, 18);
  render_text(
    label,
    (float)centered_x(label_w),
    (float)(y - 30),
    18,
    Color{200, 200, 200, 255}
  );
  DrawRectangleRounded(
    Rectangle{(float)x, (float)y, (float)width, (float)height},
    0.2f,
    8,
    ::Color{30, 30, 50, 220}
  );
  DrawRectangleRoundedLinesEx(
    Rectangle{(float)x, (float)y, (float)width, (float)height},
    0.2f,
    8,
    2.0f,
    ::Color{140, 140, 200, 255}
  );
  bool        blink_on = ((int)(GetTime() * 2) % 2) == 0;
  std::string display  = text + (blink_on ? "_" : " ");
  render_text(
    display,
    (float)(x + 12),
    (float)(y + (height - 24) / 2),
    24,
    Color{255, 255, 255, 255}
  );
}

static void update_text_input(
  const Input& input, std::string& text, size_t max_length = 16
) {
  for (char c : input.chars_typed) {
    if (text.size() < max_length) text.push_back(c);
  }
  if (key_pressed(input, KEY_BACKSPACE) && !text.empty()) text.pop_back();
}

static std::string dots() {
  int n = (int)(GetTime() * 2) % 4;
  return std::string(n, '.');
}

static bool is_super_down(const Input& input) {
#ifdef __APPLE__
  return key_down(input, KEY_LEFT_SUPER) || key_down(input, KEY_RIGHT_SUPER);
#else
  return key_down(input, KEY_LEFT_CONTROL) ||
         key_down(input, KEY_RIGHT_CONTROL);
#endif
}

Menu_Result run_menu(int window_width, int window_height, Input_Feed& inputs) {
  int W = window_width;
  int H = window_height;

  SetConfigFlags(FLAG_WINDOW_HIGHDPI);
  InitWindow(W, H, "Gods");
  SetTargetFPS(tt::TARGET_FPS);

  Menu_State state;
  int        center_y = H / 2;

  while (!WindowShouldClose()) {
    Input input = next_input(inputs);
    if (inputs.exhausted) {
      // Playback ran out of frames before the user chose a mode; bail out so
      // main() can act as if the menu was skipped (defaults to VS_AI).
      EndDrawing();
      return Menu_Result{};
    }

    // Text input for the JOINING screen.
    if (state.screen == Screen::JOINING) {
      if (is_super_down(input) && key_pressed(input, KEY_V)) {
        const char* clip = GetClipboardText();
        if (clip) state.text_input = (state.text_input + clip).substr(0, 16);
      } else {
        update_text_input(input, state.text_input);
      }
      if (key_pressed(input, KEY_ENTER) && !state.text_input.empty()) {
        state.connection = join_room(state.text_input);
        state.screen     = Screen::CONNECTING;
      }
    }

    if (state.screen == Screen::CREATING && state.connection) {
      std::string code;
      {
        std::lock_guard<std::mutex> lg(state.connection->state_lock);
        code = state.connection->room_code;
      }
      if (!code.empty() && is_super_down(input) && key_pressed(input, KEY_C)) {
        SetClipboardText(code.c_str());
      }
    }

    // Poll async connection result.
    if (state.connection) {
      if (state.connection->ready.load()) {
        Menu_Result r;
        r.mode = Menu_Result::ONLINE;
        std::lock_guard<std::mutex> lg(state.connection->state_lock);
        r.player_index      = state.connection->player_index;
        r.seed              = state.connection->seed;
        s_socket_keep_alive = state.connection->sock;
        r.online            = {
          s_socket_keep_alive.get(),
          {state.connection->friend_ip, state.connection->friend_port},
        };
        return r;
      }
      std::string err;
      {
        std::lock_guard<std::mutex> lg(state.connection->state_lock);
        err = state.connection->error;
      }
      if (!err.empty()) {
        state.error_message = err;
        state.connection.reset();
        state.screen = Screen::ONLINE;
      }
    }

    BeginDrawing();
    draw_background(input);

    if (state.screen == Screen::MAIN) {
      draw_centered_text("GODS", center_y - 180, 90, Color{255, 255, 255, 255});
      if (draw_button(input, "Play vs AI", center_y - 20)) {
        Menu_Result r;
        r.mode = Menu_Result::VS_AI;
        // Note: window stays open; main() continues using it.
        EndDrawing();
        return r;
      }
      if (draw_button(input, "Play Online", center_y + 60))
        state.screen = Screen::ONLINE;

    } else if (state.screen == Screen::ONLINE) {
      draw_centered_text("PLAY ONLINE", 150, 54, Color{255, 255, 255, 255});
      if (!state.error_message.empty()) {
        draw_centered_text(
          state.error_message, center_y - 120, 18, Color{255, 100, 100, 255}
        );
      }
      if (draw_button(input, "Create Game", center_y - 60)) {
        state.error_message.clear();
        state.connection = start_hosting();
        state.screen     = Screen::CREATING;
      }
      if (draw_button(input, "Join Game", center_y + 20)) {
        state.error_message.clear();
        state.text_input.clear();
        state.screen = Screen::JOINING;
      }
      if (draw_button(input, "Back", center_y + 120, 180, 46)) {
        state.error_message.clear();
        state.screen = Screen::MAIN;
      }

    } else if (state.screen == Screen::CREATING) {
      draw_centered_text("CREATE GAME", 150, 54, Color{255, 255, 255, 255});
      std::string code;
      if (state.connection) {
        std::lock_guard<std::mutex> lg(state.connection->state_lock);
        code = state.connection->room_code;
      }
      if (!code.empty()) {
        draw_centered_text(
          "Share this code with your friend:",
          center_y - 80,
          20,
          Color{200, 200, 200, 255}
        );
        draw_centered_text(code, center_y - 30, 50, Color{255, 215, 0, 255});
        if (draw_button(input, "Copy Code", center_y + 30, 200, 44)) {
          SetClipboardText(code.c_str());
        }
        draw_centered_text(
          "Waiting for opponent" + dots(),
          center_y + 90,
          22,
          Color{180, 180, 180, 255}
        );
      } else {
        draw_centered_text(
          "Getting your room code" + dots(),
          center_y,
          26,
          Color{200, 200, 200, 255}
        );
      }
      if (draw_button(input, "Back", center_y + 170, 180, 46)) {
        state.connection.reset();
        state.screen = Screen::ONLINE;
      }

    } else if (state.screen == Screen::JOINING) {
      draw_centered_text("JOIN GAME", 150, 54, Color{255, 255, 255, 255});
      draw_text_input("Enter room code:", state.text_input, center_y - 40);
      if (draw_button(input, "Paste", center_y + 30, 160, 44)) {
        const char* clip = GetClipboardText();
        if (clip) state.text_input = (state.text_input + clip).substr(0, 16);
      }
      if (draw_button(input, "Connect", center_y + 90) &&
          !state.text_input.empty()) {
        state.connection = join_room(state.text_input);
        state.screen     = Screen::CONNECTING;
      }
      if (draw_button(input, "Back", center_y + 160, 180, 46)) {
        state.text_input.clear();
        state.screen = Screen::ONLINE;
      }

    } else if (state.screen == Screen::CONNECTING) {
      draw_centered_text("JOIN GAME", 150, 54, Color{255, 255, 255, 255});
      draw_centered_text(
        "Connecting" + dots(), center_y - 20, 30, Color{200, 200, 200, 255}
      );
      if (draw_button(input, "Back", center_y + 100, 180, 46)) {
        state.connection.reset();
        state.screen = Screen::JOINING;
      }
    }

    EndDrawing();
  }

  // User closed window during the menu.
  CloseWindow();
  std::exit(0);
}

#endif  // !__EMSCRIPTEN__
